test_pull_requests

```diff
@@ -169,6 +203,125 @@ else
   {}
 end
 
+# Helper function to create appropriate regex for matching/extracting
+# :build_url field from bot comment
+def build_url_field_regex
+  if Properties &&
+    / \((#{Regexp::escape(Properties['jenkins_url'])}[^)]+)\)/
+  else
+    # Use this test to set the build url matching pattern
+    if ENV['BUILD_URL']
+      # Don't set build_url directly from the env var, since we want
+      # to make sure they match
+      / \((#{Regexp::escape(ENV['BUILD_URL'])})\)/
+    else
+      / \((https?:[^)]+)\)/
+    end
+  end
+end
+
+# Helper function to pull the test prefix from the settings object or
+# try to determine it from the provided bot_comment_body
+def extract_test_prefix(bot_comment_body, settings=nil)
+  if settings
+    settings['test_prefix']
+  else
+    # Do our best to find the prefix since we don't know it beforehand
+    if bot_comment_body =~ /^([^:(]+): /
+      $1.strip
+    elsif bot_comment_body =~ /^([^ (]+) /
+      $1.strip
+    end
+  end
+end
+
+# Pulls out as much metadata from the bot comment as it can, including
+# state (AKA content), Jenkins build URL, base repo commit ID, and the
+# extended tests metadata
+def extract_bot_comment_fields(bot_comment_body, settings=nil)
+  test_prefix = extract_test_prefix(bot_comment_body, settings)
+  content = nil
+  build_url = nil
+  base_commit = nil
+  extended_tests = nil
+  state = nil
+  build_url_pattern = build_url_field_regex
+  # Do our best to find the content (aka state) even if there aren't metadata in parens after it
+  if test_prefix && bot_comment_body =~ /^#{Regexp::escape(test_prefix)} ([^(]+) ?\(?/
+    content = $1.strip
+    # Find the state matching the start of the content string
+    state = CONTENT_TO_STATE.find { |k,v| content.start_with? k }[1]
+  end
+
+  if content.nil?
+    if settings.nil?
+      raise "Could not parse bot comment: '#{bot_comment_body}'"
+    end
+    raise "'#{settings['name']}' pull request not found"
+  end
+
+  build_url = bot_comment_body[build_url_pattern, 1]
+  base_commit = bot_comment_body[/ \(Base Commit: ([a-f0-9]{40})\)/, 1]
+  extended_tests = bot_comment_body[/ (\(Extended Tests: .+)/, 1]
+
+  return {
+    :prefix => test_prefix,
+    :content => content,
+    :build_url => build_url,
+    :base_commit => base_commit,
+    :extended_tests => extended_tests,
+    :state => state
+  }
+end
+
+# Helper to abstract complex waiting state check
+def waiting_any?(state)
```

Comment:
@danmcp My thinking was the function would read like: "is this waiting (on) any state?"

Maybe it would make more sense as `any_waiting?(state)` as in "is this in any waiting state?" and `queue_waiting` as in "is this in a queue waiting state?"

`waiting_state` doesn't seem to be precise enough, since we're talking about three separate states here.